src/http/matcher/mod.rs

```diff
@@ -883,3 +605,131 @@ impl<State, Body> crate::service::Matcher<State, Request<Body>> for HttpMatcherK
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use http::Version;
+
+    use crate::service::Matcher;
+
+    use super::*;
+
+    #[test]
+    fn test_matcher_ands_combination() {
+        let matcher = HttpMatcher::method_post()
+            .and(HttpMatcher::version(VersionMatcher::HTTP_2))
+            .and(HttpMatcher::domain("www.example.com"));
+        let req = Request::builder()
+            .method("POST")
+            .version(Version::HTTP_2)
+            .uri("www.example.com")
+            .body(())
+            .unwrap();
+        assert!(matcher.matches(None, &Context::default(), &req));
+    }
+
+    #[test]
+    fn test_matcher_negation_with_ands_combination() {
+        let matcher = HttpMatcher::method_post()
+            .negate()
+            .and(HttpMatcher::version(VersionMatcher::HTTP_2))
+            .and(HttpMatcher::domain("www.example.com"));
+        let req = Request::builder()
+            .method("GET")
+            .version(Version::HTTP_2)
+            .uri("www.example.com")
+            .body(())
+            .unwrap();
+        assert!(matcher.matches(None, &Context::default(), &req));
+    }
+
+    #[test]
+    fn test_matcher_ands_combination_negated() {
+        let matcher = HttpMatcher::method_post()
+            .and(HttpMatcher::version(VersionMatcher::HTTP_2))
+            .and(HttpMatcher::domain("www.example.com"))
+            .negate();
+        let req = Request::builder()
+            .method("POST")
+            .version(Version::HTTP_2)
+            .uri("www.example.com")
+            .body(())
+            .unwrap();
+        assert!(!matcher.matches(None, &Context::default(), &req));
```

Comment:
for these combo tests might be sane to verify that the logic is indeed correct (e.g. in this expect we expect to behave it like `!(A&B&C)`. Currently however this kind of test will pass even when it is `A&B&!C`.

This can be achieved by adding more request checks for the same combo matcher.